CIPs/0017.md

```diff
@@ -0,0 +1,101 @@
+
+### Upgradeability
+In order to support an upgradeable snark circuit, we utilize a `versionNumber` parameter with every proof. This corresponds to a version of the snark circuit and thus a corresponding verification key. The clients are expected to be shipped with all available verification keys, and can be upgraded to receive added verification keys.
+
+For any uses of the snark's `verify` fn in the above protocols, a client is expected to:
+* Fetch the verification key corresponding to the proof's `versionNumber` from its storage
+* The client then uses that verification key to call `verify` on the proof.
+
+### Bad proofs
+Bad proofs should be prevented from propagated throughout the network via the `verify` check that each node performs. In the case that any participant receives a proof that does not pass verification for its inputs, that participant is expected to drop the sender. 
+
+## Alternative Solutions
+
+Alternatively, `GetPlumoProofMsg` could utilize a bloom filter instead, allowing for a constant sized message. However, bloom filters allow for false positives: a false positive here would indicate to a peer that the sender has a proof when the sender does not. This could prevent certain proofs from ever getting to some nodes. For example, say node A is on the network and has received a certain amount of proofs via traditional propagation. Node A then goes down for a short period, and within that period proof P is added and propagated. Node A then comes back online, and sends a `GetPlumoProofsMsg` to request any proofs it may have missed, and sends along the corresponding bloom filter for the proofs it holds. If proof P happens to be a false positive for the set of proofs node A already has, then all of node A's peers will simply return nothing. Thus node A will never receive proof P. This can then propagate similarly through the network either through a similar path as node A, or for any nodes where all of their peers have already "lost" proof P.
+
+`GetPlumoProofMsg` could also request *all* plumo proofs from the receiver, leaving the responsibility of filtering seen proofs with the caller. This is a simpler implementation, but is less scalable as we increase the amount of stored proofs. 
+
+Specific aspects of the `les` changes may have tradeoffs that I've yet to encounter, so any input and feedback on things to consider there would be helpful.
+
+## Risks
+
+Altering the `eth` protocol introduces some risk of introducing bugs, which could inhibit a nodes ability to connect to the network.
+
+Introducing bugs to mobile client syncing. Possible effects of a faulty implementation include not being able to sync to the network, and introducing a security vulnerability which would allow the node to sync to an attack controlled chain.
```

Comment:
Please expand on this and include how it might be prevented?

Also, is there any risk related to producing snarks? Eg with the mpc setup, or if all provers are malicious

Reply:
There are potential risks inherent to the setup, but these are mitigated via crypto & opsec techniques - will need to ask the folks working on the setup for more details.

Assuming a secure setup and snark construction malicious provers' impact should be mitigated via the same tactics as malicious les servers already are. A malicious prover's proofs should not propagate throughout the network because of the verification step, and their impact on their `les` peers should be eliminated by verification checks once the light client has synced and is receiving blocks from multiple peers. This may also be mitigated earlier via multiple proofs from multiple peers, but I'm not sure on the details yet.